lib/tests/concatenation_test.cc

```diff
@@ -18,9 +19,321 @@ TEST_CASE("concatenation") {
+        REQUIRE(conf->get_int("a.y") == 42);
+    }
+
+    SECTION("concatenate two undefined substitutions with a space") {
+        auto conf = parse_config(R"(foo=abc, bar=def, a = ${foo} ${bar})")->resolve();
+        REQUIRE(conf->get_string("a") == "abc def");
+    }
+
+    SECTION("concatenate object substitutions with no space") {
+        auto conf = parse_config(R"(foo = { a : 1}, bar = { b : 2 }, x = ${foo}${bar})")->resolve();
+        REQUIRE(conf->get_int("x.a") == 1);
+        REQUIRE(conf->get_int("x.b") == 2);
+    }
+
+    SECTION("concatenate object substitutions with a space") {
+        auto conf = parse_config(R"(foo = { a : 1}, bar = { b : 2 }, x = ${foo} ${bar})")->resolve();
+        REQUIRE(conf->get_int("x.a") == 1);
+        REQUIRE(conf->get_int("x.b") == 2);
+    }
+
+    SECTION("concatenate object substitutions with quoted space") {
+        REQUIRE_THROWS_AS(
+            parse_config(R"(foo = { a : 1}, bar = { b : 2 }, x = ${foo}"  "${bar})")->resolve(),
+            config_exception
+        );
+    }
+
+    // TODO: concatSubstitutionsThatAreListsWithSpace
+
+    SECTION("concatenate list substitutions with quoted space") {
+        REQUIRE_THROWS_AS(
+            parse_config(R"(foo = [1], bar = [2], x = ${foo}"  "${bar})")->resolve(),
+            config_exception
+        );
+    }
+
+}
+
+TEST_CASE("Concatenation pending system environment lookup implementation", "[!shouldfail]") {
+    // We expect a lookup to fail and fall through in these tests.
+    // Unfortunately, because system environment lookup is not yet
+    // implemented, we fall through to that code path and then explode
+    // with an exception.
+    SECTION("string concatenation with undefined substitution") {
+        auto conf = parse_config("a = foo${?bar}")->resolve();
+        REQUIRE(conf->get_string("a") == "foo");
+    }
+
+    SECTION("object concatenation with undefined substitution") {
+        auto conf = parse_config(R"(a = { x : "foo" } ${?bar})")->resolve();
+        REQUIRE(conf->get_string("a.x") == "foo");
+    }
+
+    SECTION("concatenate two undefined substitutions") {
+        auto conf = parse_config(R"(a = ${?foo}${?bar})")->resolve();
+        REQUIRE_FALSE(conf->has_path("a"));
+    }
+
+    SECTION("concatenate several undefined substitusion") {
+        auto conf = parse_config(R"(a = ${?foo}${?bar}${?baz}${?woooo})")->resolve();
+        REQUIRE_FALSE(conf->has_path("a"));
+    }
+
+    SECTION("concatenate two undefined substitutions with a space") {
+        auto conf = parse_config(R"(a = ${?foo} ${?bar})")->resolve();
+        REQUIRE(conf->get_string("a") == " ");
+    }
+
+    SECTION("concatenate two undefined substitutions with an empty string") {
+        auto conf = parse_config(R"(a = ""${?foo}${?bar})")->resolve();
+        REQUIRE(conf->get_string("a") == "");
+    }
+}
+
+TEST_CASE("Concatenation pending '+=' implementation", "[!shouldfail]") {
+    // TODO: emptyArrayPlusEquals
```

Comment:
All += stuff fails, it's totally unimplemented. https://github.com/puppetlabs/cpp-hocon/blob/master/lib/src/config_parser.cc#L136

Probably needs Yet Another Ticket